lib/Workflow/Condition.pm

```diff
@@ -25,6 +29,140 @@ sub evaluate {
+
+    $log->is_debug
+        && $log->debug("Checking condition $condition_name");
+
+    if ( $condition_name =~ m{ \A ! }xms ) {
+
+        # this condition starts with a '!' and is thus supposed
+        # to return the opposite of an original condition, whose
+        # name is the same except for the '!'
+        $orig_condition =~ s{ \A ! }{}xms;
+        $opposite = 1;
+        $log->is_debug
+            && $log->debug(
+            "Condition starts with a !: '$condition_name'");
+    }
+
+    local $wf->{'_condition_result_cache'} =
+        $wf->{'_condition_result_cache'} || {};
+    if ( $Workflow::Condition::CACHE_RESULTS
+         && exists $wf->{'_condition_result_cache'}->{$orig_condition} ) {
+
+        # The condition has already been evaluated and the result
+        # has been cached
+        $log->is_debug
+            && $log->debug(
+            "Condition has been cached: '$orig_condition', cached result: ",
+            $wf->{'_condition_result_cache'}->{$orig_condition}
+            );
+        if ( !$opposite ) {
+            $log->is_debug
+                && $log->debug("Opposite is false.");
+            if ( !$wf->{'_condition_result_cache'}->{$orig_condition} )
+            {
+                $log->is_debug
+                    && $log->debug("Cached condition result is false.");
+            }
+            return $wf->{'_condition_result_cache'}->{$orig_condition};
+        } else {
+
+            # we have to return an error if the original cached
+            # condition did NOT fail
+            $log->is_debug
+                && $log->debug("Opposite is true.");
+            if ( $wf->{'_condition_result_cache'}->{$orig_condition} ) {
+                $log->is_debug
+                    && $log->debug("Cached condition is true.");
+                return 0;
+            }
+            return 1;
+        }
+    } else {
+
+        # we did not evaluate the condition yet, we have to do
+        # it now
+        $condition = $wf->_factory()
+            ->get_condition( $orig_condition, $wf->type );
+        $log->is_debug
+            && $log->debug( q{Evaluating condition '},
+                            $condition->name, q{'} );
+        my $return_value;
+        eval { $return_value = $condition->evaluate($wf) };
+        if ($EVAL_ERROR) {
+
+            # Check if this is a Workflow::Exception::Condition
+            if (Exception::Class->caught('Workflow::Exception::Condition')) {
+                $wf->{'_condition_result_cache'}->{$orig_condition} = 0;
+                if ( !$opposite ) {
+                    $log->is_debug
+                        && $log->debug("condition '$orig_condition' failed due to: $EVAL_ERROR");
+                    return 0;
+                } else {
+                    $log->is_debug
+                        && $log->debug("opposite condition '$orig_condition' failed due to ' . $EVAL_ERROR");
+                    return 1;
+                }
+                # unreachable
```

Comment:
heh

Reply:
If you don't mind, I'd like to leave that there: it took me quite a bit of backtracing to understand that the function would not need a closing "return" statement, because that would be dead code.